### PR TITLE
#220 Shanghai Metro Style Station Number Stage 1

### PR DESCRIPTION
**Pull request by XHG78999**
#220 First part.

### CODE REVIEW

**Comment by XHG78999** [comment]
Except of changes branch sizes in runin, I think there are nothing broken in this pull request.

**Comment by thekingofcity** [comment]
Merge latest master to make this PR mergeable.

Also, list what you have done in this PR and others need to be done in the following PR to fix #220.

**Comment by XHG78999** [comment]
> Merge latest master to make this PR mergeable.
> 
> Also, list what you have done in this PR and others need to be done in the following PR to fix #220.
> 

I will do it.

File: src/svgs/runin/runin-shmetro.tsx
Comment (thekingofcity) [inline src/svgs/runin/runin-shmetro.tsx:379]: `useMemo` can not be fed with `undefined`, learn it before use or throw memo away.

File: src/svgs/runin/runin-shmetro.tsx
Comment (thekingofcity) [inline src/svgs/runin/runin-shmetro.tsx:365]: The same as above. All 4 components share the vast majority and can be controlled with some switch line.

File: src/svgs/runin/runin-shmetro.tsx
Comment (thekingofcity) [inline src/svgs/runin/runin-shmetro.tsx:455]: The same as above. You may try {param.stn_number === true && <StationNumberPast>...</<StationNumberPast>} to git rid of large code duplication.

File: src/utils.ts
Comment (thekingofcity) [inline src/utils.ts:216]: This line need to be formatted after #232 is merged.

File: src/redux/param/reducer.ts
Comment (thekingofcity) [inline src/redux/param/reducer.ts:181]: format please

File: src/redux/param/action.ts
Comment (thekingofcity) [inline src/redux/param/action.ts:135]: The name should be updated after `stn_number` is changed.

**Review by thekingofcity — CHANGES_REQUESTED** [review]
Besides the code duplication, there is no problem with the overall approach.

Try to use several switch lines like `dy = situation === true ? -10: 10` to minimize code.

File: src/svgs/runin/runin-shmetro.tsx
Comment (XHG78999) [reply src/svgs/runin/runin-shmetro.tsx:94]: On my device , sometimes it will be undefined.

Comment (thekingofcity) [reply src/svgs/runin/runin-shmetro.tsx:94]: Can you reproduce it?

**Comment by thekingofcity** [comment]
There will be some work to do after I format the shmetro to comply with the new lint rules applied to all new PR. You may do a `git pull upstream` and fix the conflicts locally or simply format the codes with the latest eslint rules with `npm run lint:dev`.

Sorry for the inconvenience during your development.

**Review by wongchito — CHANGES_REQUESTED** [review]

File: src/constants/constants.ts
Comment (thekingofcity) [reply src/constants/constants.ts:194]: `someNewField?: string` looks good as it won't require a hard param update and can be handled with an optional chaining operator. It's up to @XHG78999 to choose his/her preferred way.

File: src/panels/design/list-shmetro.tsx
Comment (thekingofcity) [inline src/panels/design/list-shmetro.tsx:97]: By having a `<StationNumberSHMetroLi />`, I would like to extract the following to a `<PanelTypeLi/>` like gzmtr. We can use a parent component to display different styles' info_panel.

File: src/panels/design/list-shmetro.tsx
Comment (thekingofcity) [inline src/panels/design/list-shmetro.tsx:68]: I believe users will be misinformed by the UI text. When you have a text indicating `Station Number`, they may change the input field below to see the update in station number. However, the input text is used to change the line number instead of the station number. Using the existing `<Line Number />` component in gzmtr might work. You may display this component only if `showStationNumber === true`.

File: src/panels/stations/edit-diag/name-tab.tsx
Comment (thekingofcity) [inline src/panels/stations/edit-diag/name-tab.tsx:26]: `const showStationNumber = useAppSelector(store => store.param.showStationNumber);`

File: src/panels/stations/edit-diag/name-tab.tsx
Comment (thekingofcity) [inline src/panels/stations/edit-diag/name-tab.tsx:30]: `{(rmgStyle === RmgStyle.GZMTR || (rmgStyle === RmgStyle.SHMetro && stn_number)) && <NumInput {...props} />}`

File: src/panels/stations/station-list.tsx
Comment (thekingofcity) [inline src/panels/stations/station-list.tsx:79]: `const showStationNumber = useAppSelector(store => store.param.showStationNumber);`

File: src/panels/stations/station-list.tsx
Comment (thekingofcity) [inline src/panels/stations/station-list.tsx:98]: Use parenthesis correctly! You're hiding gzmtr's station number.

`            {(rmgStyle === RmgStyle.GZMTR || (rmgStyle === RmgStyle.SHMetro && showStationNumber)) && (
                <Typography className={classes.listItemNum}>{num}</Typography>
            )}`

**Review by thekingofcity — CHANGES_REQUESTED** [review]
We are almost finished except for some UI and naming tuning.

---

Besides, which way would you like to fix #220?

* Merge a single PR with multiple stages (runin, railmap, etc...).
* Merge them separately and release them to the public one by one.

I will open several subtasks for #220 if you choose the latter one.

---

Also, to which version would you like to merge? V3 or V5.

You will be responsible for merging the code to V5 if you choose to merge to `master`. This is also related to the above question as merging to V3 is required if you want this feature to be released to the public as soon as possible.

**Comment by thekingofcity** [comment]
Also, remember to resolve the conflicts.

**Review by wongchito — CHANGES_REQUESTED** [review]
Please restore the changes of `package-lock.json` as there's no dependency changes in this PR.

**Comment by XHG78999** [comment]
> Please restore the changes of `package-lock.json` as there's no dependency changes in this PR.

Ok, restored.

**Comment by thekingofcity** [comment]
Fix or comment on the reasons for unresolved conversation, please. Also, conflicts must be resolved before merging.

**Comment by XHG78999** [comment]
> Fix or comment on the reasons for unresolved conversation, please. Also, conflicts must be resolved before merging.

All conflicts in runin is about the algorithm(because of station number circle needs more size)and the code format. I am using VSCode So the code format of this and yours is not the same.

And...bulls**t china telecom, my github client cannot upload to github again because of network error. Fix them all will cost a small time.

**Comment by thekingofcity** [comment]
Besides conflicts, there are also several unresolved conversations.

**Comment by thekingofcity** [comment]
I also trigger the code check actions for this pr, try to fix these lint things.

**Comment by XHG78999** [comment]
Because of some code breaks in my local, so the PART1 and PART2 must merge in one pull request. I'll close it and reopen a new.

**Pull request closed without merging**
